### ISSUE TITLE
How to train chainer.links.EmbedID ?

### ISSUE THREAD

**Issue by sharpstill** [chainer/chainer#4454]
I also put the question on https://stackoverflow.com/questions/49207361/how-to-train-chainer-links-embedid-in-chainer
I want to train label embedding myself, (yes, label embedding like word embedding, but input is one hot vector of label)
When I found `chainer.links.EmbedID` I found example in official document, it must pass W in it.
How to train embbeding W matrix, then later we can use it to train another model?
I mean, How to train embedding vector representation of word/ label ?

**Comment by aonotas** [comment]
The input of `chainer.links.EmbedID` is following format:
```
import numpy as np
x = np.array([0, 1, 5, 8], dtype='i') # Label ID 

import chainer.links as L
n_vocab = 100
emb_dim = 50
label_embed = L.EmbedID(n_vocab, emb_dim, ignore_label=-1)

vecs = label_embed(x)
```
Please use `label ids` (not one-hot vector).
You can connect it to neural network to train loss as usual.

**Comment by sharpstill** [comment]
@aonotas Does x (in your example Label ID ) must have some order ? for instance, the order of x must have the same with article words order? Or it does not have such limit on it in learning embedding?

**Comment by sharpstill** [comment]
@aonotas  If I have two articles , each have different words number inside that article.
Is it mean we must call `vecs = label_embed(x)` multiple times?
`

**Comment by aonotas** [comment]
You can use (no limitation for the order)
```
x = np.array([8, 1, 0, 5], dtype='i') # Label ID 
```


> If I have two articles , each have different words number inside that article.
Please `concat` two variable length inputs like this:
```
x1 = np.array([0, 1, 5, 8], dtype='i') # Label ID 
x2 = np.array([5, 0, 3], dtype='i') # Label ID 

lengths = [x1.shape[0], x2.shape[0]] # 4, 3
x = np.concatenate([x1, x2], axis=0)
vecs = label_embed(x)

```
and then split concatenated vector to List with `F.split_axis`:
```
output = F.split_axis(vecs, np.cumsum(lengths)[:-1], axis=0) # List of variables
```

**Comment by sharpstill** [comment]
@aonotas If you pass the label id which does not ensure order. For example, You pass a shuffled article words, it acctually means you does not pass context information around each label/word.   How does end to end learning embedding to know this context information and learned correctly?

**Comment by sharpstill** [comment]
@aonotas If I want to pretrained a word2vec (using skip gram to train),  after trained over. Then I must load that embedding W matrix to train my second model using W matrix.  Is that mean this time I must fix the embeding matrix W to train? or just let it to end2end learning is enough? no more need addtional pretrain skip-gram word2vec step?

**Comment by aonotas** [comment]
> If you pass the label id which does not ensure order. 
> For example, You pass a shuffled article words, it acctually means you does not pass context information around each label/word. 
> How does end to end learning embedding to know this context information and learned correctly?

I can't understand what you mean in English. Please give some example code? 


> How to set pretrained word2vec?
```
pretrained_w2v = np.array(...) # load word2vec vectors as numpy format
label_embed = L.EmbedID(n_vocab, emb_dim, ignore_label=-1)

# set pretrained vectors
label_embed.W.data[:] = pretrained_w2v[:]
```

**Comment by sharpstill** [comment]
@aonotas I know this is to load pretrained word2vec parameters, I am not very experience in this area, I worked on vision field. I want to know is it **better** to fix `label_embeded`  layer W after load pretrained or not to fix parameter update in training?

First question, I mean , You says: "no limitation for the order" , in this situation, The article words are in chaos order sequence, How does `L.EmbedID` to train the word embedding in such chaos article?

**Comment by aonotas** [comment]
@sharpstill 
If you solve your problems of L.EmbedID, please close this issue.
